File: 6. Queue & Stack/279. Perfect Squares/squares.cpp
class Solution
{
public:
    int ans = INT_MAX;
    void dfs(int n, int sq, int sum)
    {
        if (n == 0)
        {
            ans = min(ans, sum);
            return;
        }
        if (sq == 0)
        {
            return;
        }
        int square = sq * sq;

        if (n >= square)
        {
            dfs(n % square, sq - 1, sum + n / square);
        }
        dfs(n, sq - 1, sum);
    }

    int numSquares(int n)
    {
        dfs(n, sqrt(n), 0);
        return ans;
    }
};

File: 6. Queue & Stack/279. Perfect Squares/squares.py
class Solution:
    def dfs(self, n, sq, _sum):
        if n == 0:
            self.ans = min(self.ans, _sum)
            return
        if sq == 0:
            return
        
        square = sq * sq

        if n >= square:
            self.dfs(n % square, sq - 1, _sum + floor(n/square))
        
        self.dfs(n, sq-1, _sum)

    def numSquares(self, n: int) -> int:
        self.ans = n
        self.dfs(n, floor(sqrt(n)), 0)
        return self.ans

